src/Pact/Typechecker.hs

```diff
@@ -370,21 +371,55 @@ substAppDefun :: Maybe (TcId, AST Node) -> Visitor TC Node
 substAppDefun (Just (defArg,appAst)) Pre t@Var {..}
   | defArg == _aId _aNode = assocAST defArg appAst >> return appAst
   | otherwise = return t
-substAppDefun Nothing Post (App appNode appFun appArgs) = do -- Post, to allow AST subs first
-    af <- case appFun of
-      FNative {} -> return appFun -- noop
-      FDefun {..} -> do
-        -- assemble substitutions, and also associate fun ty with app args
-        let mangledFty = mangleFunType (_aId appNode) _fType
-        subArgs <- forM (zip3 _fArgs appArgs (_ftArgs mangledFty)) $ \(fa,aa,_ft) -> -- do
-          -- assocAstTy (_aNode aa) (_aType ft) -- this was causing recursive var refs, bailing for now
-          return (_aId (_nnNamed fa),aa)
-        -- associate fun return ty with app
-        assocAstTy appNode (_ftReturn mangledFty)
-        let subDefArg b fa = walkAST (substAppDefun (Just fa)) b
-        fb' <- forM _fBody $ \bAst -> foldM subDefArg bAst subArgs
-        return $ set fBody fb' appFun
-    return (App appNode af appArgs)
+substAppDefun Nothing Post (App appNode fun args) = do -- Post, to allow AST subs first
+    fun' <- case fun of
+      FNative {} -> return fun -- noop
+      FDefun {_fType,_fArgs,_fBody} -> do
+        let FunType argTys retTy = mangleFunType (_aId appNode) _fType
```

Comment:
can we break this out into it's own function, ie the `FDefun` branch? It's gotten quite long

Reply:
Will do